Content.Client/Nyanotrasen/CartridgeLoader/Cartridges/GlimmerMonitorUiFragment.xaml.cs

```diff
@@ -0,0 +1,123 @@
+        RobustXamlLoader.Load(this);
+        IoCManager.InjectDependencies(this);
+
+        Orientation = LayoutOrientation.Vertical;
+        HorizontalExpand = true;
+        VerticalExpand = true;
+
+        var intervalGroup = new ButtonGroup();
+        IntervalButton1.Group = intervalGroup;
+        IntervalButton5.Group = intervalGroup;
+        IntervalButton10.Group = intervalGroup;
+
+        IntervalButton1.Pressed = true;
+
+        IntervalButton1.OnPressed += _ => UpdateState(_cachedValues);
+        IntervalButton5.OnPressed += _ => UpdateState(_cachedValues);
+        IntervalButton10.OnPressed += _ => UpdateState(_cachedValues);
+
+        SyncButton.OnPressed += _ => OnSync?.Invoke(true);
+    }
+
+    public void UpdateState(List<int> glimmerValues)
+    {
+        _cachedValues = glimmerValues;
+        if (glimmerValues.Count < 1)
+            return;
+
+        MonitorBox.RemoveAllChildren();
+
+        var glimmerLabel = new Label();
+        glimmerLabel.Text = Loc.GetString("glimmer-monitor-current-glimmer", ("glimmer", glimmerValues[^1]));
+        MonitorBox.AddChild(glimmerLabel);
+
+        var formattedValues = FormatGlimmerValues(glimmerValues);
+        var graph = new GlimmerGraph(_resourceCache, formattedValues);
+        graph.SetSize = (450, 250);
+        MonitorBox.AddChild(graph);
+    }
+
+
+    private List<int> FormatGlimmerValues(List<int> glimmerValues)
+    {
+        var returnList = glimmerValues;
+
+        if (IntervalButton5.Pressed)
+        {
+            returnList = GetAveragedList(glimmerValues, 5);
+        } else if (IntervalButton10.Pressed)
+        {
+            returnList = GetAveragedList(glimmerValues, 10);
+        }
+
+        return ClipToFifteen(returnList);
+    }
+
+    /// <summary>
+    /// Format glimmer values to get <=15 data points correctly.
+    /// </summary>
+    private List<int> ClipToFifteen(List<int> glimmerValues)
+    {
+        List<int> returnList;
+
+        if (glimmerValues.Count <= 15)
+        {
+            returnList = glimmerValues;
+        }
+        else
+        {
+            returnList = glimmerValues.Skip(glimmerValues.Count - 15).ToList();
+        }
+
+        return returnList;
+    }
+
+    private List<int> GetAveragedList(List<int> glimmerValues, int interval)
+    {
+        var returnList = new List<int>();
+
+        int i = 0;
+        while (i < glimmerValues.Count - interval - 1)
+        {
+            var tempList = glimmerValues.GetRange(i, interval);
+
+            int total = 0;
+            foreach (var value in tempList)
+            {
+                total += value;
+            }
+
+            returnList.Add(total / interval);
+            i += interval;
+        }
+
+        int remainingTotal = 0;
+        foreach (var value in glimmerValues.Skip(i))
+        {
+            remainingTotal += value;
+        }
+
+        returnList.Add(remainingTotal / glimmerValues.Skip(i).Count());
+
+        return returnList;
```

Comment:
```suggestion
    private List<int> GetAveragedList(IEnumerable<int> glimmerValues, int interval)
    {
        var returnList = new List<int>();

        var subtotal = 0;
        var elementsPassed = 0;
        for (int i = 0; i < glimmerValues.Count(); ++i)
        {
            subtotal += glimmerValues.ElementAt(i);

            ++elementsPassed;
            if (elementsPassed == interval)
            {
                returnList.Add(subtotal / interval);
                subtotal = 0;
                elementsPassed = 0;
            }
        }

        if (elementsPassed != 0)
            returnList.Add(subtotal / elementsPassed);

        return returnList;
    }
```

Let me know if you like this rewrite. I only tested it in a local console to make sure it works. It functions differently at interval boundaries (off by one) but otherwise seems to line up with your original function without creating new lists for every slice.

Reply:
It works fine, and we can use all the client perf we can get.